src/hotspot/share/runtime/javaThread.cpp

```diff
@@ -1015,6 +1016,38 @@ void JavaThread::set_exception_oop(oop o) {
   Atomic::store(&_exception_oop, o);
 }
 
+// Effectively wait for any deoptimization to complete and clear _monitor_chunk.
+class ReadMonitorChunksHandshake : public HandshakeClosure {
+  MonitorChunk* _mc;
+ public:
+  ReadMonitorChunksHandshake(): HandshakeClosure("ReadMonitorChunks") {}
+  void do_thread(Thread* thr) {
+    JavaThread* target = JavaThread::cast(thr);
+    _mc = target->monitor_chunks();
+  }
+  MonitorChunk* monitor_chunks() {
+    return _mc;
+  }
+};
+
+// _monitor_chunks is set and cleared by deoptimization.
+// If non-null, deoptimization is in progress and the value
+// will become junk soon.  Waiting for deopt is the only action.
+// A ThreadLocalHandshake will mean deopt is complete.
+// If at a Safepoint, the value can be read but it is surely nullptr.
+MonitorChunk* JavaThread::monitor_chunks_safe() const {
+  MonitorChunk* chunks = nullptr;
+  if (SafepointSynchronize::is_at_safepoint() || this == Thread::current()) {
+    chunks = _monitor_chunks; // will be null, as deopt frees when finished
+  } else {
+    ReadMonitorChunksHandshake rmch;
```

Comment:
How expensive is the Handshake versus the direct read?  Is it worth optimistically reading using `monitor_chunks()` and only attempt the handshake if it returns non-null?

Or is there an API to probe if a thread is in a deopt that we can wrap around the handshake?

If the handshake is cheap enough, then this isn't worth looking at further.

Reply:
Actually yes I was thinking about checking the value for non-null first, I will try that out.  No handshake has to be faster... 8-)

Reply:
Updated - if _monitor_chunks is null, avoid the handshake, and the check if we are at a safepoint.